pkg/services/consumer.go

```diff
@@ -62,8 +65,27 @@ func (s *sqlConsumerService) Replace(ctx context.Context, consumer *api.Consumer
 	return consumer, nil
 }
 
+// Delete will remove the consumer from the storage. Currently, it will:
+// 1. Perform a hard delete on the consumer, the resource creation will be blocked after it.
+// 2. Forbid consumer deletion if there are associated resources.
+// 3. The deleting resources(marked as deleted) will still block the consumer deletion.
+// TODO: Additional deletion options or strategies may be added in the future.
 func (s *sqlConsumerService) Delete(ctx context.Context, id string) *errors.ServiceError {
-	if err := s.consumerDao.Delete(ctx, id); err != nil {
+	consumer, err := s.Get(ctx, id)
+	if err != nil {
+		return err
+	}
+	_, e := s.resourceDao.FirstByConsumerName(ctx, consumer.Name, true)
+	if e == nil {
+		return errors.Forbidden("Resources associated with the consumer: %s", consumer.Name)
+	}
+
+	if e != gorm.ErrRecordNotFound {
+		return errors.GeneralError("Unable to get resources by consumer: %s", e)
+	}
+
+	// e is record not found
+	if err := s.consumerDao.Delete(ctx, id, true); err != nil {
 		return handleDeleteError("Consumer", errors.GeneralError("Unable to delete consumer: %s", err))
```

Comment:
maybe add service error type at https://github.com/openshift-online/maestro/blob/main/pkg/errors/errors.go#L78-L90 and use it for consumer deletion

Reply:
As we discussed, we can leverage the forbidden type directly.